[Step4] 자동차 경주

kimseonghun opened the pull request:
step 4 구현 내용 PR 보냅니다~
이전 step3에서 주신 피드백 내용 반영도 포함돼있습니다!

이번에도 역시 꼬리 재귀 공부 목적으로 우승자 계산 로직에 적용해보고자 했으나 오히려 복잡도가 늘어나는 것 같아
`generateSequence()`를 활용해봤습니다.

이번에는 변경 사항이 크게 없어 (물론 커밋이 14개라는게 조금 어폐가 있긴하지만...) 코멘트 달 내용이 마땅히 없네요.

감사합니다 🙇

kimseonghun committed:
[step4] docs: step4 기능 요구 사항 정리

kimseonghun committed:
[step4] feat: 자동차 입력을 받을 수 있도록 입력 인터페이스에 함수 추가

kimseonghun committed:
[step4] feat: Name VO 추가

kimseonghun committed:
[step4] feat: RacingCar 객체에 Name VO 의존 추가

kimseonghun committed:
[step4] feat: RacingGame 자동차 이름 문자열로 객체를 생성할 수 있도록 secondary constructor 추가

- 문자열을 인자로 받아야하기에 입력 인터페이스 명세 수정 (함수 반환값: List<String> -> String)

kimseonghun committed:
[step4] feat: RacingCarController에서 새로운 함수로 갈음

- 자동차 이름을 입력받는 입력 명세 사용
- 자용차 이름 문자열을 인자로 받는 RacingGame 객체 생성자 사용
- 위 작업으로 필요없는 코드 제거

kimseonghun committed:
[step4] feat: 출력 구현 로직 변경

- 자동차 이름도 같이 출력할 수 있도록 DTO 객체에 필드 추가

kimseonghun committed:
[step4] chore: 구현 완료 기능 항목 체크

kimseonghun committed:
[step4] feat: 우승자 계산 로직 추가

kimseonghun committed:
[step4] refactor: 확장함수 모음 파일 추가

kimseonghun committed:
[step4] feat: 우승자 결과 출력 인터페이스 명세 추가

kimseonghun committed:
[step4] feat: Controller에서 우승자 계산 및 출력 호출 로직 추가

kimseonghun committed:
[step4] refactor: when 연산자에 (true) 제거

kimseonghun committed:
[step3] refactor: 코드 리뷰 반영

- 매직넘버 상수화
- 파라미터 이름 변경

Flamme1004K commented on README.md:
```diff
@@ -1,9 +1,14 @@
 - [x] 자동차를 시도할 횟수만큼 이동시킨다.
 - [x] 자동차의 상태를 화면에 출력한다.
+
+## step 4 기능 요구 사항 정리
+- [x] 자동차 대수를 숫자 대신 이름으로 입력받는다. (,기준 이름이 나뉘며 이름은 5자를 넘을 수 없다.)
+- [x] 이동하는 자동차를 출력할 때 자동차 이름을 같이 출력한다.
+- [x] 자동차 경주 게임을 완료한 후 누가 우승했는지를 알려준다. 우승자는 한 명 이상일 수 있다.
```
요구사항 정리 👍

Flamme1004K commented on src/main/kotlin/nexstep/mission/racingcar/RacingCar.kt:
```diff
@@ -1,8 +1,14 @@
-class RacingCar(var position: Int = 0) {
+class RacingCar(
+    val name: Name,
+    var position: Int = 0
```
가변인 position을 public으로 한다면 바깥에서도 수정 가능하지 않을까요~? 🤔

kimseonghun replied:
오 그렇네요, 반영해놓겠습니다!

하나 궁금한게 있는데요,
보통 이런 경우 코틀린 진영에선 어떤 방식으로 문제를 풀어나가나요?

이 문제를 저는 `position` 필드에 private 접근제어자를 추가해주고 별도의 getter 함수를 추가하는 방향으로 변경했는데요,
아래와 같이 풀어나갈까도 고민했었습니다.
```kotlin
class RacingCar(val name) {
   var position = 0
      private set
}
```
위와 같이 하지 않은 이유는 primary constructor에 필드를 둚으로써 해당 객체의 생성 시 사용성을 높이기 위함이에요. 필드 초기화 권한을 사용하는 측에 위임해서 더 안전하고 효율적으로 객체를 사용하게 만들 수 있을거라 생각했어요~

하지만 한편으론 너무 Java 스럽게 문제를 해결한게 아닌가 라는 의구심도 드네요 😂

Flamme1004K replied:
좋은 의견 감사합니다! 

두 가지 방안이 있어보입니다.

1. 매개변수를 통하여 프로퍼티를 주입하는 방법, 이렇게 하면 기본 생성자로써의 목적도 가능하고 또한 초기화 후에 getter/setter에 대해서 접근제어가 가능해요.

``` kotlin
class RacingCar(
    val name: Name,
    position: Int = 0,
) {

    var position: Int = position
        private set
```

2.  backing property를 사용하는 방법. 생성자에는 private 한 var를 선언하고, 이를 꺼내올 때는 불변으로 꺼내오는 방법. 

```kotlin 
class RacingCar(
    val name: Name,
    private var _position: Int,
) {

    val position
        get() = _position
```

저는 두 가지 방법 중에, 요즘은 1번을 많이 사용하는 중입니다. 그 이유는 2번을 사용하게 되면 `Named Argument` 사용 시에 `_`을 붙여서 할당해야하기 떄문에 클라이언트 측에서 사용하때 불편함이 있을 수도 있기 때문입니다. 😊

kimseonghun replied:
오 그렇군요 저도 첫번째 내용이 좋아보이네요! 감사합니다~

근데 혹시 첫번째 주신 의견에서 primary constructor의 position 필드에 val을 선언하는게 맞죠?

kimseonghun replied:
아 이게 val을 선언하면 안되는군요... 조금 헷갈리네요 이걸 val을 선언하지 않으면 생성자 자리에 있는 position은 클래스 내부 필드가 되는게 아닌가보네요...?

Flamme1004K replied:
금일 강의를 보고 의문이 풀리셨을 것이라고 생각합니다 😊

Flamme1004K commented on src/main/kotlin/nexstep/mission/racingcar/RacingGame.kt:
```diff
@@ -1,21 +1,36 @@
 
+import nexstep.mission.extenstionfun.pop
+
+fun List<RacingCar>.toSortedMutable() = this.sortedByDescending { it.position }.toMutableList()
```
불변 리스트를 가변 리스트로 변경한 이유를 알고 싶습니다!

kimseonghun replied:
nexstep.mission.extenstionfun.pop 함수를 사용하기 위해서인데요,
우승자 계산(`winner()`) 시 `pop()` 함수로 첫번째 값을 꺼내고 확인하고를 반복하기 위함이어요!

Flamme1004K commented on src/main/kotlin/nexstep/mission/racingcar/RacingGame.kt:
```diff
@@ -1,21 +1,36 @@
+    fun winner(favorites: MutableList<RacingCar> = this.racingCars.toSortedMutable()): List<RacingCar> =
+        generateSequence(favorites.pop()) { next(it, favorites) }.toList()
+
+    private fun next(previous: RacingCar, favorites: MutableList<RacingCar>) =
```
리턴 객체도 같이 표시해주는걸 추천드립니다!

실무에서 한번 겪었던 일인데요. 리턴 객체를 표기 안 해주다가 Unit으로 바꾼 함수가 있었는데요. 찾기가 엄청 골치 아프더라구요. 😭

kimseonghun replied:
아하 그렇군요 반영해놓을게요!

Flamme1004K commented on src/main/kotlin/nexstep/mission/racingcar/vo/Name.kt:
```diff
@@ -0,0 +1,7 @@
+
+data class Name(val value: String) {
+    init {
+        require(this.value.length <= 5) { "자동차 이름은 5자를 초과할 수 없습니다." }
```
require 사용 좋습니다 👍👍

Flamme1004K commented on src/test/kotlin/nexstep/mission/racingcar/RacingGameTest.kt:
```diff
@@ -21,12 +23,27 @@ class RacingGameTest : FunSpec({
         }
     }
+
+    test("경주 자동차의 position값이 제일 높은 우승자를 계산한다.") {
```
우승자 테스트 👍👍 우승자의 이름까지 테스트하면 더 정확하지 않을까요?

kimseonghun replied:
좋아요~!

Flamme1004K commented on src/test/kotlin/nexstep/mission/racingcar/vo/NameTest.kt:
```diff
@@ -0,0 +1,24 @@
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+
+class NameTest : FunSpec({
```
생성자 테스트 👍

Flamme1004K reviewed (changes requested):
안녕하세요! 

전체적으로 꼼꼼히 구현을 해주셨군요~👍

몇 가지 코멘트를 남겼으니 확인 부탁드리겠습니다!

확인 후 다시 리뷰 요청 드리며, 궁금하신 점에 대해서는 DM 부탁드립니다!

마지막으로 미션 화이팅입니다! 🙌

kimseonghun committed:
[step4] chore: 주석 내용 정리

kimseonghun committed:
[step4] refactor: 리뷰반영

- 함수 반환값 명시
- 테스트 코드 수정

kimseonghun committed:
[step4] refactor: MutableList -> List 변경

- https://github.com/next-step/kotlin-racingcar/pull/788#discussion_r1017868919

kimseonghun committed:
[step4] refactor: RacingCar의 position 필드의 접근 제어자를 private로 변경

kimseonghun committed:
[step4] refactor: 테스트명 수정

kimseonghun commented:
형준님 안녕하세요~
남겨주신 피드백 반영해서 push 했습니다!

그 다음 step5를 진행하려고 했는데요, 요구사항을 보니 이미 현재 step4 에 모두 반영된 것들로 보여서요~
step5 진행을 어떻게 하면 좋을지 모르겠네요. 스킵해도 무관한지 궁금합니다.

Flamme1004K commented on src/test/kotlin/nexstep/mission/racingcar/RacingGameTest.kt:
```diff
@@ -1,32 +1,54 @@
+
+        result.forAll {
+            it.position() shouldBe 5
+            it.name.value shouldBe "pobi"
```
만약 우승자가 2명인 경우에는 테스트를 어떻게 짜야할까요? 🤔

kimseonghun replied:
그러게요 이 부분때문에 기존 테스트 코드의 given의 객체들을 수정하긴 했었는데 다른 방법이 있나보네요?
한번 찾아볼게요!

Flamme1004K reviewed (approved):
안녕하세요!

피드백 반영 감사합니다. 💯

그동안 step4 고생하셨습니다! 다음 단계를 위하여 merge 하도록 하겠습니다 🙌

코멘트를 남겼으니, step5 자동차 경주 진행하며, 같이 반영해주시면 되겠습니다.

궁금하신 점에 대해서는 DM 부탁드리겠습니다. 🙇🏻

다음 미션도 화이팅입니다! 🤩

Flamme1004K commented:
step5는 3가지 방안이 있어보여요 😊

1. 리팩토링 해보고 싶은 부분이 있다면 리팩토링
2.  TDD로 처음부터 끝까지 다시 구현
3.  요번 미션을 그만하고 싶으시다면 DM 부탁드립니다!

kimseonghun commented:
의견 감사합니다!
그렇다면 저는 1번을 선택할게요!
이번 PR에서 주신 피드백 반영 및 추가로 한번 정리 해보고 싶은 부분을 정리하는 스텝으로 가져가볼겠습니다 :)
